### PR TITLE
[RUNTIME] Add function name to PackedFunc and TypedPackedFunc

### PR DESCRIPTION
**Pull request by tkonolige**
This PR adds a `name` field to both `PackedFunc` and `TypedPackedFunc`. This name defaults to `<anonymous>`, but I've also added code that sets the name for `TVM_REGISTER_GLOBAL().set_body()` and `TVM_REGISTER_GLOBAL().set_body_typed()`. With this change, the error messages for incorrect number of arguments is now a lot more clear: it prints the function name in the error message.

I'm not sure if the approach I am using here makes sense. I'd like feedback on whether or not this is the correct direction to go.

@tqchen @jroesch @junrushao1994

### CODE REVIEW

**Comment by junrushao1994** [comment]
Adding function name to packed functions definitely make sense to me. I wonder if we can go a bit further:
* On C++ side, enhance `TVM_REGISTER_GLOBAL` to set the name of the packed func
* On python side, enhance `register_func` to set of the name of the packed func too
* What about the rust side?

**Comment by tkonolige** [comment]
This PR changes `TVM_REGISTER_GLOBAL` so that is sets the name of the packed func. I can do the same on the python side.

**Comment by junrushao1994** [comment]
Shall we accomplish this more automatically by modifying the `TVM_REGISTER_GLOBAL` macro itself?

**Comment by junrushao1994** [comment]
In file `runtime/registry.h`, Right now the macro is:

```C++
#define TVM_REGISTER_GLOBAL(OpName) \
  TVM_STR_CONCAT(TVM_FUNC_REG_VAR_DEF, __COUNTER__) = ::tvm::runtime::Registry::Register(OpName)
```

I think we can change the `runtime::Registry::set_body` method, adding `f.name_ = ...` to it

**Comment by tkonolige** [comment]
That's what I do here: https://github.com/apache/tvm/pull/7108/files#diff-0fe12f71cc9718f4bf446609180e9c13ca870f0a1ecf080f330b7af8743a0d43R61

**Comment by junrushao1994** [comment]
Oh nice! I didn't notice that :-)

File: include/tvm/runtime/packed_func.h [hunk 1]
```diff
@@ -293,6 +304,8 @@ class TypedPackedFunc<R(Args...)> {
   friend class TVMRetValue;
   /*! \brief The internal packed function */
   PackedFunc packed_;
+  /*! \brief The name identifying this function */
+  String name_;
```
Comment (zhiics) [inline include/tvm/runtime/packed_func.h:308]: should we just use the `name_` of packed_?

Comment (tkonolige) [reply include/tvm/runtime/packed_func.h]: Good point, I've made this change.

**Comment by zhiics** [comment]
BTW, can you add a unit test? I think you can add it to the existing gtest.

**Comment by tqchen** [comment]
Thanks @tkonolige . First of all, I think it is very useful to be able to support additional meta data like names when the user can provid eit.

On the other hand, attaching a String that defaults to a concrete value would mean additional overhead for PackedFunc creation, and may be prohibitive if we want to create PackedFunc on the fly(with a goal to gain std::function level speed).

It would be good put more thoughts into this and think about how to improve  the impl in general. As a first step, it certainly does not hurt to make the name Optional and default to nullptr, which reduces the string construction cost. 

We can alternatively think about the future path of bringing in PackedFunc as a sub-class of Object, and potentially sub-class PackedFunc. So the default one remains un-named, while there is a named subclass.

Given it is quite a core change, an RFC would be great.

**Review by tqchen — CHANGES_REQUESTED** [review]
see the comments, thanks for proposing the change

**Comment by tqchen** [comment]
Thanks @Tristan Konolige for the packed func naming PR. 
It was on my mind for a while but didn't get to think about all the design tradeoffs. An RFC that quickly polls dev community would be great, specifically there are a few design spectrums(feel free to include these in the rfc):

Naming convention:
N0: name_hint: might be  better name since it does not uniquely identify things
N1: name

What is the default string:
S0: anonymous
S1: use a special sub-class of PackedFunc(need to move PackedFunc to object system)
S2: always attach Optional<String> since the ref object is relatively cheap, default to nullptr

API choices to set the name
A0: Allow mutation, add setter API
A1: Only allow immutable construction, might affect how we construct from C caallback
A2: force to use name_hint, allow idempotent setting, if the name_hint is nullptr, allow set, otherwise allow set only if the new name is same as the old one. Error out if mismatch.

**Comment by tqchen** [comment]
closing in favor of the alternative capturing approach

**Pull request closed without merging**